### ISSUE TITLE
ReportUnnecessaryComparison bogus error

### ISSUE THREAD

**Issue by DiegoBaldassarMilleuno** [microsoft/pyright#7431]
**Describe the bug**
The code below reports:
`error: Condition will always evaluate to False since the types "Literal[1]" and "Literal[0]" have no overlap (reportUnnecessaryComparison)`.
This is not correct, since the print statement is actually run.

**Code or Screenshots**
```python
# pyright: strict

def test(nr_cycles: int) -> None:
    progress_nr = 0
    def done_callback() -> None:
        nonlocal progress_nr
        progress_nr += 1
        if progress_nr % 10 == 0:
            print(f'{progress_nr}...')
    
    for _ in range(nr_cycles):
        done_callback()

if __name__ == '__main__':
    test(100)
```

**VS Code extension or command-line**
Running from command line as `LANG=C pyright filename.py`.
Version reported is `pyright 1.1.353`.
OS is MacOS Sonoma 14.2.1, running Python 3.12.1 from Homebrew.
I believe this bug has been introduced recently, though I can't be 100% sure.

**Comment by erictraut** [comment]
Thanks for the bug report. The problem here isn't the `reportUnnecessaryComparisonError`. That's just a symptom of a different bug. Pyright is incorrectly evaluating the type of `progress_nr` as `Literal[1]` when it should be `int` in this case. The type should be widened to `int` because the `+= 1` operation is performed within a separate execution context from the one that initially assigns the value `0` to the variable. I'll work on a fix.

**Comment by erictraut** [comment]
This will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.354, which I just published.
